File: go/crypto-square/crypto_square.go
package cryptosquare

import (
	"math"
	"regexp"
	"strings"
)

// NormalizeText normalizes text.
func NormalizeText(s string) (res string) {
	s = strings.ToLower(s)
	re := regexp.MustCompile("[0-9A-Za-z]")
	for _, r := range s {
		if re.MatchString(string(r)) {
			res += string(r)
		}
	}
	return
}

// Size returns the column and row lengths.
func Size(s string) (c int, r int) {
	c = int(math.Ceil(math.Sqrt(float64(len(s)))))
	r = int(math.Floor(math.Sqrt(float64(len(s)))))
	if c*r < len(s) {
		r++
	}
	return
}

// Rows breaks up a string into chunks.
func Rows(s string, size int) (rows []string) {
	temp := ""
	for i, r := range s {
		if i != 0 && i%size == 0 {
			rows = append(rows, temp)
			temp = ""
		}
		temp += string(r)
	}
	rows = append(rows, temp)
	return
}

// CodedMessage translates rows into a string.
func CodedMessage(rows []string, size int) (res string) {
	for i := 0; i < size; i++ {
		for j := 0; j < len(rows); j++ {

			if i < len(rows[j]) {
				res += string(rows[j][i])
			} else {
				res += " "
			}
		}
	}
	return
}

// Output breaks up the coded message.
func Output(s string, size int) (res string) {
	for i := 0; i < len(s); i++ {
		if i != 0 && i%size == 0 {
			res += " "
		}
		res += string(s[i])
	}
	return
}

// Encode encodes a string.
func Encode(s string) string {
	normalizedText := NormalizeText(s)
	c, r := Size(normalizedText)
	rows := Rows(normalizedText, c)
	codedMessage := CodedMessage(rows, c)
	return Output(codedMessage, r)
}

File: python/crypto-square/crypto_square.py
import math

def normalize_text(text):
  return ''.join([c for c in list(text.lower()) if c.isalnum()])

def encode(plain_text):
  normalized_text = normalize_text(plain_text)
  segment_length = int(math.ceil(math.sqrt(len(normalized_text))))

  segments = []
  normalized_segment = ''
  for i, c in enumerate(normalized_text):
    normalized_segment += c
    if i % segment_length == segment_length - 1:
      segments.append(normalized_segment)
      normalized_segment = ''
  if len(normalized_segment) > 0:
    segments.append(normalized_segment)

  cipher_text = []
  for char_pos in range(segment_length):
    cipher_segment = ''
    for i, segment in enumerate(segments):
      cipher_segment += segment[char_pos] if len(segment) > char_pos else ' '
    cipher_text.append(cipher_segment)
  return ' '.join(cipher_text)